### PR TITLE
(SERVER-2679) Ensure return callback runs before actual return

### PR DESCRIPTION
**Pull request by mwaggett**
This commit puts the `instance-returned` callback before the actual `return`
again in the public-facting `return-to-pool` method. Previously, if a request
was waiting for a new instance, calling `return` first could trigger a new
borrow so quickly that the borrow event callback would trigger before the return
event callback. For the instance pool, because instances ids are reused, instead
of 1. removing the instance from the `borrowed-instances` metric and then 2.
adding it back with the new borrow event, we ended up 1. updating the
`borrowed-instances` with the new event and then 2. removing the instance from
`borrowed-instances`, which caused the metric to lose track of a whole instance.

### CODE REVIEW

**Comment by mwaggett** [comment]
confirmed with https://github.com/puppetlabs/puppetserver/pull/2264 that the failures I was seeing are now gone 🎉

File: src/clj/puppetlabs/services/jruby_pool_manager/impl/reference_pool.clj [hunk 1]
```diff
@@ -131,7 +133,7 @@
           (jruby-agents/send-agent modify-instance-agent
                                    #(flush-if-at-max-borrows pool-context instance)))
         ;; Return the worker-id, to be used in metrics and event logging
-        (.getId (Thread/currentThread)))))
+        (pool-protocol/worker-id pool-context instance))))
```
Comment (Magisus) [inline src/clj/puppetlabs/services/jruby_pool_manager/impl/reference_pool.clj]: I think we might not need to do this anymore, if you're calling `worker-id` explicitly in the return event registering?

Comment (mwaggett) [reply src/clj/puppetlabs/services/jruby_pool_manager/impl/reference_pool.clj]: This is helpful in testing that the worker-id is the same at borrow-time and return-time.

Though the real reason I kept it was to change as little as possible.. 😁

Comment (Magisus) [reply src/clj/puppetlabs/services/jruby_pool_manager/impl/reference_pool.clj]: Okay I think it's fine for now then.

File: src/clj/puppetlabs/services/protocols/jruby_pool.clj [hunk 2]
```diff
@@ -23,6 +23,10 @@
     [pool-context]
     "Unlocks the JRuby pool, allowing borrows to proceed.")
 
+  (worker-id
+    [pool-context instance]
+    "Returns a reference to a worker id (instance id or thread id).")
```
Comment (Magisus) [inline src/clj/puppetlabs/services/protocols/jruby_pool.clj]: Can you rephrase this a little to say "Returns the worker id for a given instance"? It's not really a reference we're returning, and since that word is used so much with the new pool, I'd rather not overload it.

File: src/clj/puppetlabs/services/protocols/jruby_pool.clj [hunk 3]
```diff
@@ -23,6 +23,10 @@
     [pool-context]
     "Unlocks the JRuby pool, allowing borrows to proceed.")
 
+  (worker-id
+    [pool-context instance]
+    "Returns the worker id for given instance (instance id or thread id).")
```
Comment (mwaggett) [inline src/clj/puppetlabs/services/protocols/jruby_pool.clj]: @Magisus updated
